cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -79,7 +81,7 @@ fn identity_for_exchange(cred: &ClusterCredential) -> Result<Identity> {
 /// The rust derive macro together with the kube macro creates serializable and deserializable
 /// resources based on the struct. See https://docs.rs/kube/0.43.0/kube/ for more details.
 #[derive(CustomResource, Deserialize, Serialize, Clone, Debug)]
-#[kube(group = "login.concierge.pinniped.dev", version = "v1alpha1", kind = "TokenCredentialRequest", namespaced)]
```

Comment:
Did you try with 0.5.0 without this `namespaced` option here? I am surprised that it would work without it (in that I assume this macro derives either a namespaced or non-namespaced resource). Ah, or maybe it just determines whether the metadata *must* have a namespace set to be valid or something.

Reply:
I wanted to try, but I don't know why I've been facing the same issue of the missing cookie (#-2505). My requests never reached the pinniped proxy... and I'm using the same files as I was using the last week...

Reply:
(just dumping here the logs, later will edit the comment) 

` kubectl -n kube-system logs  -l component=kube-apiserver  -f`

```
E0311 11:51:45.229972       1 controller.go:116] loading OpenAPI spec for "v1alpha1.login.concierge.pinniped.dev" failed with: OpenAPI spec does not exist
I0311 11:51:45.229994       1 controller.go:129] OpenAPI AggregationController: action for item v1alpha1.login.concierge.pinniped.dev: Rate Limited Requeue.
E0311 11:51:51.340896       1 authentication.go:53] Unable to authenticate the request due to an error: invalid bearer token
```

Reply:
> (just dumping here the logs, later will edit the comment)
> 
> ` kubectl -n kube-system logs -l component=kube-apiserver -f`
> 
> ```
> E0311 11:51:45.229972       1 controller.go:116] loading OpenAPI spec for "v1alpha1.login.concierge.pinniped.dev" failed with: OpenAPI spec does not exist
> I0311 11:51:45.229994       1 controller.go:129] OpenAPI AggregationController: action for item v1alpha1.login.concierge.pinniped.dev: Rate Limited Requeue.
> E0311 11:51:51.340896       1 authentication.go:53] Unable to authenticate the request due to an error: invalid bearer token
> ```
> 
> Took 7 min to pull image from projects.registry.vmware.com
> 

Yeah, I found the same - took ages here (although I did suspend in between, perhaps causing problems).

And yes, I see the same error in my cluster api server logs (note you can just export them in kind too, which is sometimes nicer, with `kind export logs --name kubeapps` - gives you all logs in a tmp dir). This error is just because pinniped-concierge apparently isn't defining an OpenAPI spec (which apiservices are meant to do, apparently). ([Here's an example of a similar issue for prometheus operator](https://github.com/prometheus-operator/kube-prometheus/issues/304))

> Pinniped pod
> 
> ```
> E0311 12:48:25.794626       1 reflector.go:138] /cache/gomodcache/k8s.io/client-o@v0.20.1/tools/cache/reflector.go:167: Failed to watch *v1beta1.FlowSchema: failed to list *v1beta1.FlowSchema: flowschemas.flowcontrol.apiserver.k8s.io is forbidden: User "system:serviceaccount:pinniped-concierge:pinniped-concierge" cannot list resource "flowschemas" in API group "flowcontrol.apiserver.k8s.io" at the cluster scope
> ```

Also harmless, I see this too (in my working env). I guess this is because `flowschemas` are a part of the [API priority and fairness feature which is only available in k8s v1.20](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/) and my Kind cluster is currently using 1.18 (yes I should update). But that's just a guess.

> 
> Bad certificate, I copypasted bad
> 
> ```
> E0311 13:09:04.884712       1 oidc.go:224] oidc authenticator: initializing plugin: Get "https://172.19.0.2:32000/.well-known/openid-configuration": x509: certificate signed by unknown authority
> E0311 13:09:12.731575       1 runtime.go:76] Observed a panic: runtime error: invalid memory address or nil pointer dereference
> k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1.1(0xc0009d7080)
>         /cache/gomodcache/k8s.io/apiserver@v0.20.1/pkg/server/filters/timeout.go:106 +0x113 panic(0x1bd5380, 0x2d74ec0)
> ```

Do you mean that you fixed the above because it was a copy-paste error?

> 
> Namespaced authenticator in 0.6.0,
> 
> ```
> I0311 13:27:15.625383       1 certs_observer.go:66] certsObserverController Sync updated certs in the dynamic cert provider
> I0311 13:27:15.625745       1 certs_expirer.go:88] certsExpirerController Sync found a renew delta of -599h49m34.374266991s
> E0311 13:27:27.132579       1 runtime.go:76] Observed a panic: runtime error: invalid memory address or nil pointer dereference
> ```

Do you mean the above are logs from a pinniped-concierge 0.6.0 pod? Unsure about this one as I see the following in my 0.6.0 pod logs:

```
I0312 04:01:27.976314       1 certs_observer.go:66] certsObserverController Sync updated certs in the dynamic cert provider
I0312 04:01:28.061193       1 certs_expirer.go:88] certsExpirerController Sync found a renew delta of -599h59m48.938814425s
I0312 04:01:28.061526       1 certs_manager.go:136] certsManagerController Sync successfully created secret
```

> 
> Recreate the entire cluster again
> 
> ```
>    1 clientconn.go:948] ClientConn switching balancer to "pick_first"
> E0311 16:00:39.151159       1 controller.go:116] loading OpenAPI spec for "v1alpha1.login.concierge.pinniped.dev" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body: error trying to reach service: dial tcp 10.96.251.99:443: connect: connection refused
> , Header: map[Content-Type:[text/plain; charset=utf-8] X-Content-Type-Options:[nosniff]]
> I0311 16:00:39.151178       1 controller.go:129] OpenAPI AggregationController: action for item v1alpha1.login.concierge.pinniped.dev: Rate Limited Requeue.
> ...
> E0311 16:23:39.636478       1 controller.go:116] loading OpenAPI spec for "v1alpha1.login.concierge.pinniped.dev" failed with: OpenAPI spec does not exist
> I0311 16:23:39.636517       1 controller.go:129] OpenAPI AggregationController: action for item v1alpha1.login.concierge.pinniped.dev: Rate Limited Requeue.
> ```

Also unrelated, I think? The first one is just the apiserver failing to even reach the `v1alpha1.login.concierge.pinniped.dev` apiservice at first, because the pods wouldn't have been running (you can check how k8s views the api service with `kubectl get apiservices` and see there whether it thinks it's running or not). And the second one is just the same as earlier - the pinniped apiservice apparently doesn't include an OpenAPI spec yet.

Reply:
> (just dumping here the logs, later will edit the comment)

Whoops, I forgot to update the comments with more descriptive messages! Sorry about that!

> And yes, I see the same error in my cluster api server logs. This error is just because pinniped-concierge apparently isn't defining an OpenAPI spec (which apiservices are meant to do, apparently). ([Here's an example of a similar issue for prometheus operator](https://github.com/prometheus-operator/kube-prometheus/issues/304))

Ahh, ok, I see. I thought the resources weren't being properly installed at all. Thanks for the clarification here.

> (note you can just export them in kind too, which is sometimes nicer, with `kind export logs --name kubeapps` - gives you all logs in a tmp dir). 

Not aware of this feature. Thanks!

> Also harmless, I see this too (in my working env). I guess this is because `flowschemas` are a part of the [API priority and fairness feature which is only available in k8s v1.20](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/) and my Kind cluster is currently using 1.18 (yes I should update). But that's just a guess.

I use the latest Kind version with k8s v1.20.2, but I agree, it should be harmless.

> Do you mean that you fixed the above because it was a copy-paste error?

Yes, I copied  the `client-certificate-data` instead of `certificate-authority-data` :disappointed:  

> Do you mean the above are logs from a pinniped-concierge 0.6.0 pod? Unsure about this one as I see the following in my 0.6.0 pod logs:

Yes, they are; from time to time I continue having some panics even if it is working now.

> Also unrelated, I think? The first one is just the apiserver failing to even reach the `v1alpha1.login.concierge.pinniped.dev` apiservice at first, because the pods wouldn't have been running (you can check how k8s views the api service with `kubectl get apiservices` and see there whether it thinks it's running or not). 

Yep, you're right. Thanks for the `kubectl get apiservices` command, really useful.

So, finally, the problem was an OSI Layer 8 one :P  I was omitting one step, copying the cert from the additional cluster in the Kubeapps values. I noticed it after your changes in the PR, so thanks a lot for the help!